bindings/java/java/src/main/java/org/ethereum/evmc/EvmcVm.java

```diff
@@ -3,41 +3,79 @@
 // Licensed under the Apache License, Version 2.0.
 package org.ethereum.evmc;
 
-import org.ethereum.evmc.EvmcLoaderException;
-
+import java.io.IOException;
 import java.nio.ByteBuffer;
-import java.util.Objects;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
 
 /**
  * The Java interface to the evm instance.
  *
  * <p>Defines the Java methods capable of accessing the evm implementation.
  */
 public final class EvmcVm implements AutoCloseable {
-  private static EvmcVm evmcVm;
-  private static boolean isEvmcLibraryLoaded = false;
+  private static final Throwable errorLoadingEvmc;
   private ByteBuffer nativeVm;
+
+  static {
+    Throwable error = null;
+    try {
+      // load so containing the jni bindings to evmc
+      System.loadLibrary("libevmc-java");
+    } catch (UnsatisfiedLinkError e) {
+      String extension = null;
+      String os = System.getProperty("os.name").toLowerCase();
+      if (os.contains("win")) {
+        extension = "dll";
+      } else if (os.contains("nix") || os.contains("nux") || os.contains("aix")) {
+        extension = "so";
+      } else if (os.contains("mac")) {
+        extension = "dylib";
+      } else {
+        error = e;
+      }
+      if (extension != null) {
+        try {
+          Path evmcLib = Files.createTempFile("libevmc-java", extension);
+          Files.copy(
+              EvmcVm.class.getResourceAsStream("/libevmc-java." + extension),
```

Comment:
I'm not sure what is the best practice for this, but can you use fixed extension for all OSes? E.g. put it in the JAR as `java/src/main/resources/libevmc-java.jni` and the load as `System.load("/tmp/libevmc-java.jni")`. You will not have to figure out the OS-specific DLL extension.

Did you test it on Windows? On Windows the DLLs don't have the "lib" prefix.

Reply:
You would want to package all native libraries in the Java jar. This means the jar contains a .so, a .dylib, and a .dll, at a minimum. You can always let folks place the native lib in the path if they want to load their custom library or use a 32 bit system.

I guess the best way to test on Windows would be to have a CircleCI Windows build, I don't have a Windows machine.